File: typescript/src/cost/ledger.ts
import type { TraceSpan } from "../tracing/tracer.js";

/**
 * Per-run cost ledger. `RunBudget` caps token *count*; the profiler shows a live dollar gauge;
 * this is the after-the-fact accounting in between - it turns the `costUsd` every model_call span
 * already carries into a total plus a breakdown by model and by agent role, so "what did this run
 * cost, and where?" is one function call, not a spreadsheet. It reads the trace the run already
 * emitted, so it adds no instrumentation and no extra model calls (notes section 8, 11).
 */
export interface CostLine {
  calls: number;
  inputTokens: number;
  outputTokens: number;
  costUsd: number;
}

export interface CostSummary {
  total: CostLine;
  byModel: Record<string, CostLine>;
  byAgent: Record<string, CostLine>;
}

const emptyLine = (): CostLine => ({ calls: 0, inputTokens: 0, outputTokens: 0, costUsd: 0 });

function add(line: CostLine, span: TraceSpan): void {
  line.calls += 1;
  line.inputTokens += span.tokenUsage?.inputTokens ?? 0;
  line.outputTokens += span.tokenUsage?.outputTokens ?? 0;
  line.costUsd += span.costUsd ?? 0;
}

/** Reduce a run's spans to a cost summary. Only `model_call` spans carry spend. */
export function summarizeCost(spans: readonly TraceSpan[]): CostSummary {
  const summary: CostSummary = { total: emptyLine(), byModel: {}, byAgent: {} };
  for (const span of spans) {
    if (span.kind !== "model_call") continue;
    const model = (span.attributes?.["model"] as string | undefined) ?? "unknown";
    const agent = span.agentRole ?? "unknown";
    add(summary.total, span);
    add((summary.byModel[model] ??= emptyLine()), span);
    add((summary.byAgent[agent] ??= emptyLine()), span);
  }
  return summary;
}

/** A compact, human-readable rendering of a cost summary for a CLI or a log line. */
export function formatCostSummary(summary: CostSummary): string {
  const usd = (n: number) => `$${n.toFixed(6)}`;
  const line = (label: string, l: CostLine) => `  ${label.padEnd(24)} ${String(l.calls).padStart(3)} calls  ${String(l.inputTokens + l.outputTokens).padStart(8)} tok  ${usd(l.costUsd)}`;
  const rows: string[] = [`Run cost: ${usd(summary.total.costUsd)} over ${summary.total.calls} model call(s)`];
  rows.push("by model:");
  for (const [model, l] of Object.entries(summary.byModel)) rows.push(line(model, l));
  rows.push("by agent:");
  for (const [agent, l] of Object.entries(summary.byAgent)) rows.push(line(agent, l));
  return rows.join("\n");
}

File: python/src/multiagent_boilerplate/cost/ledger.py
"""Per-run cost ledger. RunBudget caps token *count*; the profiler shows a live dollar gauge;
this is the after-the-fact accounting in between - it turns the cost_usd every model_call span
already carries into a total plus a breakdown by model and by agent role, so "what did this run
cost, and where?" is one function call. It reads the trace the run already emitted, so it adds no
instrumentation and no extra model calls (notes section 8, 11). Mirrors typescript/src/cost/ledger.ts."""

from __future__ import annotations

from dataclasses import dataclass, field

from ..tracing.tracer import TraceSpan


@dataclass
class CostLine:
    calls: int = 0
    input_tokens: int = 0
    output_tokens: int = 0
    cost_usd: float = 0.0


@dataclass
class CostSummary:
    total: CostLine = field(default_factory=CostLine)
    by_model: dict[str, CostLine] = field(default_factory=dict)
    by_agent: dict[str, CostLine] = field(default_factory=dict)


def _add(line: CostLine, span: TraceSpan) -> None:
    usage = span.token_usage or {}
    line.calls += 1
    line.input_tokens += usage.get("inputTokens", 0)
    line.output_tokens += usage.get("outputTokens", 0)
    line.cost_usd += span.cost_usd or 0.0


def summarize_cost(spans: list[TraceSpan]) -> CostSummary:
    """Reduce a run's spans to a cost summary. Only `model_call` spans carry spend."""
    summary = CostSummary()
    for span in spans:
        if span.kind != "model_call":
            continue
        model = (span.attributes or {}).get("model", "unknown")
        agent = span.agent_role or "unknown"
        _add(summary.total, span)
        _add(summary.by_model.setdefault(model, CostLine()), span)
        _add(summary.by_agent.setdefault(agent, CostLine()), span)
    return summary


def format_cost_summary(summary: CostSummary) -> str:
    """A compact, human-readable rendering of a cost summary for a CLI or a log line."""

    def line(label: str, ln: CostLine) -> str:
        tok = ln.input_tokens + ln.output_tokens
        return f"  {label:<24} {ln.calls:>3} calls  {tok:>8} tok  ${ln.cost_usd:.6f}"

    rows = [f"Run cost: ${summary.total.cost_usd:.6f} over {summary.total.calls} model call(s)", "by model:"]
    rows += [line(model, ln) for model, ln in summary.by_model.items()]
    rows.append("by agent:")
    rows += [line(agent, ln) for agent, ln in summary.by_agent.items()]
    return "\n".join(rows)
